ansible_base/rbac/permission_registry.py

```diff
@@ -0,0 +1,155 @@
+            # NOTE: right now this only supports names that are the same as the type
+            if parent_field_name == parent_model_name:
+                if model_name in seen:
+                    continue
+                seen.add(model_name)
+                child_model = self._name_to_model[model_name]
+                child_filters.append((parent_field_name, child_model))
+                for next_parent_filter, grandchild_model in self.get_child_models(child_model, seen=seen):
+                    child_filters.append((f'{next_parent_filter}__{parent_field_name}', grandchild_model))
+        return child_filters
+
+    def call_when_apps_ready(self, apps, app_config):
+        from ansible_base.rbac import triggers
+        from ansible_base.rbac.evaluations import bound_has_obj_perm, bound_singleton_permissions, connect_rbac_methods
+
+        self.apps = apps
+        self.apps_ready = True
+
+        if self.team_model not in self._registry:
+            self._registry.add(self.team_model)
+
+        post_migrate.connect(triggers.post_migration_rbac_setup, sender=app_config)
+
+        self.user_model.add_to_class('has_obj_perm', bound_has_obj_perm)
+        self.user_model.add_to_class('singleton_permissions', bound_singleton_permissions)
+        post_delete.connect(triggers.rbac_post_user_delete, sender=self.user_model, dispatch_uid='permission-registry-user-delete')
+
+        # Temporary HACK until the created_by and modified_by cascade behavior is resolved
+        def clear_created_assignments(instance, *args, **kwargs):
+            from ansible_base.rbac.models import RoleTeamAssignment, RoleUserAssignment
+
+            RoleUserAssignment.objects.filter(created_by=instance).update(created_by=None)
+            RoleTeamAssignment.objects.filter(created_by=instance).update(created_by=None)
+
+        pre_delete.connect(clear_created_assignments, sender=self.user_model, dispatch_uid='permission-registry-clear-created-assignments')
+
+        # end HACK
+
+        for cls in self._registry:
+            triggers.connect_rbac_signals(cls)
+            connect_rbac_methods(cls)
+
+        for cls, relationship, role_name in self._tracked_relationships:
+            if role_name in self._trackers:
+                tracker = self._trackers[role_name]
+            else:
+                tracker = triggers.TrackedRelationship(cls, role_name)
+                self._trackers[role_name] = tracker
+            tracker.initialize(relationship)
+
+    @cached_property
+    def team_model(self):
+        return self.apps.get_model(settings.ANSIBLE_BASE_TEAM_MODEL)
+
+    @cached_property
+    def team_ct_id(self):
+        return self.content_type_model.objects.get_for_model(self.team_model).id
+
+    @cached_property
+    def user_model(self):
+        return get_user_model()
+
+    @cached_property
+    def content_type_model(self):
+        return self.apps.get_model('contenttypes.ContentType')
```

Comment:
I agree, it shouldn't be cached.

ContentType is only like this for future reasons. The permission class is declared as a swappable model right now. That means that a project points a setting to the model to use, and that model has to meet certain minimum criteria. That could be the `auth.Permission` model or a custom model. The reason why I have ContentType handled in a similar way is because I intend to make it swappable for https://github.com/ansible/django-ansible-base/issues/80. We probably _can_ merge the permission models, but the ContentType will need to swap depending on if it's a plain app, or a controller of external permissions.

Reply:
You can't durably add a `ContentType` entry for a model that doesn't exist in the local server. That's what I have in mind. The model will exist in a _different_ server, but permissions for those objects will be tracked locally. That's not a problem for the `object_id` of a GenericForeign key because it's not relational in the first place, lol. But it doesn't work with the stock `ContentType` model.